[SPARK-20140][DStream] Remove hardcoded kinesis retry wait and max retries

yssharma opened the pull request:
## What changes were proposed in this pull request?

The pull requests proposes to remove the hardcoded values for Amazon Kinesis - MIN_RETRY_WAIT_TIME_MS, MAX_RETRIES.

This change is critical for kinesis checkpoint recovery when the kinesis backed rdd is huge.
Following happens in a typical kinesis recovery :
- kinesis throttles large number of requests while recovering
- retries in case of throttling are not able to recover due to the small wait period
- kinesis throttles per second, the wait period should be configurable for recovery

The patch picks the spark kinesis configs from:
- spark.streaming.kinesis.retry.wait.time
- spark.streaming.kinesis.retry.max.attempts

Jira : https://issues.apache.org/jira/browse/SPARK-20140

## How was this patch tested?

Modified the KinesisBackedBlockRDDSuite.scala to run kinesis tests with the modified configurations. Wasn't able to test the patch with actual throttling.

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented:
Hi @shaynativ, this PR looks a non-trivial change that needs a JIRA. Please refer http://spark.apache.org/contributing.html.

HyukjinKwon commented:
Hi @yssharma, this PR looks a non-trivial change that needs a JIRA. Please refer http://spark.apache.org/contributing.html.

yssharma commented:
Have got this JIRA ticket for the patch - https://issues.apache.org/jira/browse/SPARK-20140

yssharma commented:
Oh right you meant the PR title format. I will reject this pr and post a new one. Thanks @HyukjinKwon

HyukjinKwon commented:
You could just edit the title. I think closing this and opening new one is also fine and an option.

yssharma commented:
@HyukjinKwon What should be the next steps for this PR. Are there any Spark-Kinesis experts who can review the patch ?

HyukjinKwon commented:
I am not used to Kinesis. I usually click blame button and check both the recent code modifier and committer, e.g., https://github.com/yssharma/spark/blame/4b589adeaef540f6227266ecc628ad41ef0733c3/external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisBackedBlockRDD.scala

I think I assume @tdas and @brkyvz are experts in this area.

yssharma commented:
You're a gen @HyukjinKwon 💯 . I will wait for Tathagata and Burak's inputs then :)

yssharma commented:
@tdas @brkyvz 
- Added new changes for removing other constants hardcoded at multiple places.
- Squashed 3 commits into single commit

yssharma commented:
Can I get some feedback here please @tdas @brkyvz Thanks :)

SparkQA commented:
**[Test build #3634 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3634/testReport)** for PR 17467 at commit [`ccb6c19`](https://github.com/apache/spark/commit/ccb6c19dfaec4b7667552dfeed3da730923b7b64).

SparkQA commented:
**[Test build #3634 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3634/testReport)** for PR 17467 at commit [`ccb6c19`](https://github.com/apache/spark/commit/ccb6c19dfaec4b7667552dfeed3da730923b7b64).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

yssharma commented:
@srowen - Could I get some love here as well. Thanks

srowen reviewed (commented):
It seems reasonable but I don't know enough to review Kinesis changes, really

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisBackedBlockRDD.scala:
```diff
@@ -251,21 +255,22 @@ class KinesisSequenceRangeIterator(
-    var waitTimeMs = MIN_RETRY_WAIT_TIME_MS
+    var kinesisWaitTimeMs =
+      kinesisConfigs.getOrElse("spark.streaming.kinesis.retry.wait.time", "100").toInt
+    val kinesisMaxRetries =
```
define this once during class initialization

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisBackedBlockRDD.scala:
```diff
@@ -251,21 +255,22 @@ class KinesisSequenceRangeIterator(
     var retryCount = 0
-    var waitTimeMs = MIN_RETRY_WAIT_TIME_MS
+    var kinesisWaitTimeMs =
+      kinesisConfigs.getOrElse("spark.streaming.kinesis.retry.wait.time", "100").toInt
```
define this once during class initialization as well

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisBackedBlockRDD.scala:
```diff
@@ -292,9 +297,3 @@ class KinesisSequenceRangeIterator(
 }
-
-private[streaming]
-object KinesisSequenceRangeIterator {
```
keep the default values here please

brkyvz commented on docs/streaming-kinesis-integration.md:
```diff
@@ -216,3 +216,7 @@ de-aggregate records during consumption.
   - `InitialPositionInStream.TRIM_HORIZON` may lead to duplicate processing of records where the impact is dependent on checkpoint frequency and processing idempotency.
+
+- Kinesis retry configurations
+ - `spark.streaming.kinesis.retry.wait.time` : Config for wait time between Kinesis retries (in milliseconds). Default is 100 ms.
```
We can take duration strings, such as `100ms` or `1s`, I would prefer we use that.

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisBackedBlockRDD.scala:
```diff
@@ -83,7 +83,8 @@ class KinesisBackedBlockRDD[T: ClassTag](
     val messageHandler: Record => T = KinesisInputDStream.defaultMessageHandler _,
-    val kinesisCreds: SparkAWSCredentials = DefaultCredentials
+    val kinesisCreds: SparkAWSCredentials = DefaultCredentials,
+    val kinesisConf: Map[String, String] = Map.empty
```
do you need this to be provided as a `Map`? You already have `sc` being passed in with all the configurations

budde replied:
+1. I think reading the config values from ```sc``` will be a much cleaner approach.

yssharma replied:
+1

brkyvz reviewed (changes requested):
Thanks for working on this PR! Sorry for not being able to take a look earlier. Some feedback:

1. I thought the `kinesisConfigs` map was there to make testing easier, but it didn't since you set the SparkConfs as well as the map, so I suggest we remove it
2. Let's make `spark.streaming.kinesis.retry.wait.time` a duration string, allow it to take `100ms` instead of have it be in millis. You can use https://github.com/apache/spark/blob/master/common/network-common/src/main/java/org/apache/spark/network/util/JavaUtils.java#L244
3. May be a tad bit shorter to name as:
`spark.streaming.kinesis.retry.waitTime`
`spark.streaming.kinesis.retry.maxAttempts`

yssharma commented:
@brkyvz - thanks for taking time to review the patch. appreciate it.
Implemented all your suggestions. Now passing a new map for the kinesis configs and added mechanism to use the builder for the configs.

As for the spark context, I wanted to use the sparkcontext available in `KinesisBackedBlockRDD` directly as well (instead of creating a new config map), but the sc in `KinesisBackedBlockRDD`
is not available, and trying to use it there causes serialization errors. Passing a different config map looked like the only simple solution to access the kineses configs.

The patch now doesnot use the `sc` at all and expects a kinesisConf to be passed to the `KinesisInputDStream` builder directly.

Let me know your thoughts. Thanks again for the review comments.

yssharma commented:
Just for info, while trying to use the `sc` in the `KinesisBackedBlockRDD ` : 

`- Basic reading from Kinesis *** FAILED ***
  org.apache.spark.SparkException: Task not serializable
  at org.apache.spark.util.ClosureCleaner$.ensureSerializable(ClosureCleaner.scala:298)
  at org.apache.spark.util.ClosureCleaner$.org$apache$spark$util$ClosureCleaner$$clean(ClosureCleaner.scala:288)
  at org.apache.spark.util.ClosureCleaner$.clean(ClosureCleaner.scala:108)
  at org.apache.spark.SparkContext.clean(SparkContext.scala:2284)
  at org.apache.spark.SparkContext.runJob(SparkContext.scala:2058)
  at org.apache.spark.SparkContext.runJob(SparkContext.scala:2084)
  at org.apache.spark.rdd.RDD$$anonfun$collect$1.apply(RDD.scala:936)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
  at org.apache.spark.rdd.RDD.withScope(RDD.scala:362)
  ...
  Cause: java.io.NotSerializableException: org.apache.spark.SparkContext
Serialization stack:
	- object not serializable (class: org.apache.spark.SparkContext, value: org.apache.spark.SparkContext@60c1663c)
	- field (class: org.apache.spark.streaming.kinesis.KinesisBackedBlockRDD, name: org$apache$spark$streaming$kinesis$KinesisBackedBlockRDD$$sc, type: class org.apache.spark.SparkContext)
	- object (class org.apache.spark.streaming.kinesis.KinesisBackedBlockRDD, KinesisBackedBlockRDD[0] at BlockRDD at KinesisBackedBlockRDD.scala:90)
	- field (class: org.apache.spark.NarrowDependency, name: _rdd, type: class org.apache.spark.rdd.RDD)
	- object (class org.apache.spark.OneToOneDependency, org.apache.spark.OneToOneDependency@52a33c3f)
	- writeObject data (class: scala.collection.immutable.List$SerializationProxy)
	- object (class scala.collection.immutable.List$SerializationProxy, scala.collection.immutable.List$SerializationProxy@71ed560f)
	- writeReplace data (class: scala.collection.immutable.List$SerializationProxy)
	- object (class scala.collection.immutable.$colon$colon, List(org.apache.spark.OneToOneDependency@52a33c3f))
	- field (class: org.apache.spark.rdd.RDD, name: org$apache$spark$rdd$RDD$$dependencies_, type: interface scala.collection.Seq)
	- object (class org.apache.spark.rdd.MapPartitionsRDD, MapPartitionsRDD[1] at map at KinesisBackedBlockRDDSuite.scala:83)
	- field (class: org.apache.spark.rdd.RDD$$anonfun$collect$1, name: $outer, type: class org.apache.spark.rdd.RDD)
	- object (class org.apache.spark.rdd.RDD$$anonfun$collect$1, <function0>)
	- field (class: org.apache.spark.rdd.RDD$$anonfun$collect$1$$anonfun$13, name: $outer, type: class org.apache.spark.rdd.RDD$$anonfun$collect$1)
	- object (class org.apache.spark.rdd.RDD$$anonfun$collect$1$$anonfun$13, <function1>)
  at org.apache.spark.serializer.SerializationDebugger$.improveException(SerializationDebugger.scala:40)
  at org.apache.spark.serializer.JavaSerializationStream.writeObject(JavaSerializer.scala:46)
  at org.apache.spark.serializer.JavaSerializerInstance.serialize(JavaSerializer.scala:100)
  at org.apache.spark.util.ClosureCleaner$.ensureSerializable(ClosureCleaner.scala:295)
  at org.apache.spark.util.ClosureCleaner$.org$apache$spark$util$ClosureCleaner$$clean(ClosureCleaner.scala:288)
  at org.apache.spark.util.ClosureCleaner$.clean(ClosureCleaner.scala:108)
  at org.apache.spark.SparkContext.clean(SparkContext.scala:2284)
  at org.apache.spark.SparkContext.runJob(SparkContext.scala:2058)
  at org.apache.spark.SparkContext.runJob(SparkContext.scala:2084)
  at org.apache.spark.rdd.RDD$$anonfun$collect$1.apply(RDD.scala:936)`

yssharma commented:
Waiting for review @brkyvz . Thanks.

budde commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisBackedBlockRDD.scala:
```diff
@@ -147,6 +152,14 @@ class KinesisSequenceRangeIterator(
   private var internalIterator: Iterator[Record] = null
 
+  // variable for kinesis wait time interval between next retry
+  private var kinesisWaitTimeMs = JavaUtils.timeStringAsMs(
```
Can't this be a ```val```?

yssharma replied:
No, this value is modified after waits -
`kinesisWaitTimeMs *= 2  // if you have waited, then double wait time for next round`

budde replied:
I don't think you want to do this-- ```kinesisWaitTimeMs``` is never reset to the default value after the retry loop exists. I think you should make this a ```val``` and introduce a ```var``` initialized to its value within ```retryOrTimeout()``` to store the wait time for each retry iteration.

budde commented:
Not a Spark committer, but I've contributed to this component in the past. I would strongly prefer an approach that avoids adding an additional parameter to all of the Kinesis classes if the ```SparkConf``` from ```sc``` can be used instead.  I haven't looked at the initial version of your code, but based on the stacktrace you've posted it seems like you might've been referencing ```sc``` directly from code running running on the executor (e.g. the ```KinesisSequenceRangeIterator``` instance created in the ```compute()``` method via ```getBlockFromKinesis()```). Did you try simply extracting the two config values from ```sc``` and storing them as fields for ```KinesisBackedBlockRDDPartition``` then passing them as constructor arguments to ```KinesisSequenceRangeIterator```?

budde commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisInputDStream.scala:
```diff
@@ -249,6 +252,17 @@ object KinesisInputDStream {
     }
 
     /**
+      * Sets the [[SparkAWSCredentials]] to use for authenticating to the AWS CloudWatch
```
Documentation is not correct

budde commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisInputDStream.scala:
```diff
@@ -249,6 +252,17 @@ object KinesisInputDStream {
+      *
+      * @param conf: Map[String, String] to use for CloudWatch authentication
+      */
+    def kinesisConf(conf: Map[String, String]): Builder = {
```
If using ```SparkConf``` to store these custom config values doesn't end up being feasible then I'd strongly prefer that we follow the existing approach and have separate builder methods for setting the retry wait time and max attempts.

yssharma replied:
Do you think it would be better to pass values to builder rather than a map of configs. I thought map of configs can be easily extended when we need to support new configurations without code changes?
What is your thought on values+builder per config vs one map for all configs ?

budde replied:
If you want the extensibility of a key/value map for configs then I would go the route of getting a solution that uses ```SparkConf``` to do that in order to use the existing facilities provided by Spark. It doesn't make sense to me to introduce a key/value map just for Kinesis, especially since the naming of your keys (e.g. ```spark.streaming.kinesis.retry.waitTime```) would indicate to most users that these are ```SparkConf``` params, not a Kinesis-specific mapping that must be manually set up and passed to the Kinesis stream builder.

yssharma commented:
@budde - thanks for taking time to review it. Appreciate it.

yssharma commented:
@budde @brkyvz - Implemented the review changes. Please review.

- Using SparkConf for all the user parameters
- removed kinesisWait to be val instead of var
- fixed documentation

budde commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisBackedBlockRDD.scala:
```diff
@@ -147,6 +153,17 @@ class KinesisSequenceRangeIterator(
 
+  // variable for kinesis wait time interval between next retry
+  private val kinesisWaitTimeMs = JavaUtils.timeStringAsMs(
+    Try {sparkConf.get("spark.streaming.kinesis.retry.waitTime")}
```
It may also be useful to declare these keys as public constants in a sensible location such as the [companion object to ```KinesisInputDStream```](https://github.com/apache/spark/blob/master/external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisInputDStream.scala#L84), e.g.:

```scala
object KinesisInputDStream {
...
  /**
   * Relevant doc
   */
  val RETRY_WAIT_TIME_KEY = "spark.streaming.kinesis.retry.waitTime"
 
  /**
   * Relevant doc
   */
  val RETRY_MAX_ATTEMPTS_KEY = "spark.streaming.kinesis.retry.maxAttempts"
...
```

This will make things a little less brittle for users who want to dynamically fill in SparkConf values in their apps. You would also be able use these constants in unit tests here.

budde commented on external/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisBackedBlockRDDSuite.scala:
```diff
@@ -101,6 +101,37 @@ abstract class KinesisBackedBlockRDDTests(aggregateTestData: Boolean)
+      testUtils.endpointUrl, fakeBlockIds(allRanges.size),
+      allRanges.map { range => SequenceNumberRanges(Array(range)) }.toArray,
+      sparkConf = sc.getConf
+    ).map { bytes => new String(bytes).toInt }.collectPartitions()
```
*nit:* move this to the end of previous line

budde reviewed (commented):
Is there any way we can test the actual behavior being added here, e.g. add unit tests to check the configured wait time and max attempts are being respected?

yssharma commented:
@budde : Implemented review changes and checked scala code style checks

yssharma commented:
@budde - Not sure if we can exactly test the configured timeouts. I have debugged the flow in my testcase to verify that the custom values are passed down to the Kinesis backed block rdd.

budde commented:
@yssharma Fair enough. I'll try to get your update reviewed later today

budde commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisBackedBlockRDD.scala:
```diff
@@ -295,6 +306,23 @@ class KinesisSequenceRangeIterator(
-  val MAX_RETRIES = 3
-  val MIN_RETRY_WAIT_TIME_MS = 100
+  /**
+   * The maximum number of attempts to be made to kinesis. Defaults to 3.
```
*nit:* **K**inesis

budde commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisBackedBlockRDD.scala:
```diff
@@ -295,6 +306,23 @@ class KinesisSequenceRangeIterator(
+  val MAX_RETRIES = "3"
+
+  /**
+   * The interval between consequent kinesis retries. Defaults to 100ms.
```
*nit:* **K**inesis

budde commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisBackedBlockRDD.scala:
```diff
@@ -295,6 +306,23 @@ class KinesisSequenceRangeIterator(
+  val MIN_RETRY_WAIT_TIME_MS = "100ms"
+
+  /**
+   * Key for configuring the retry wait time for kinesis. The values can be passed to SparkConf.
```
*nit:* I'd make the following tweaks here:

```scala
/**
 * SparkConf key for configuring the wait time to use before retrying a Kinesis attempt.
 */
```

budde commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisBackedBlockRDD.scala:
```diff
@@ -295,6 +306,23 @@ class KinesisSequenceRangeIterator(
+  val RETRY_WAIT_TIME_KEY = "spark.streaming.kinesis.retry.waitTime"
+
+  /**
+   * Key for configuring the number of retries for kinesis. The values can be passed to SparkConf.
```
*nit:* I'd make the following tweaks here:

```scala
/**
 * SparkConf key for configuring the maximum number of retries used when attempting a Kinesis
 * request.
 */
```

budde commented on docs/streaming-kinesis-integration.md:
```diff
@@ -216,3 +216,7 @@ de-aggregate records during consumption.
   - `InitialPositionInStream.LATEST` could lead to missed records if data is added to the stream while no input DStreams are running (and no checkpoint info is being stored).
   - `InitialPositionInStream.TRIM_HORIZON` may lead to duplicate processing of records where the impact is dependent on checkpoint frequency and processing idempotency.
+
+- Kinesis retry configurations
```
@brkyvz or another Spark committer might have better suggestions here, but I would suggest making this section a new heading (rather than part of **Kinesis Checkpointing**) and adding a brief explanatory sentence, e.g.:

```
#### Kinesis retry configuration
- A Kinesis DStream will retry any failed request to the Kinesis API. The following SparkConf properties can be set in order to customize the behavior of the retry logic:
```

followed by the rest of your changes here.

This also reminds me that I owe @brkyvz a change to add docs for the stream builder interface here :)

budde reviewed (commented):
A few more small nits but otherwise this is looking quite good!

felixcheung commented:
@brkyvz are you ok with this PR at a high level? If yes, I could help with review and shepherd this

yssharma commented:
@budde - Added changes for the minor review comments and docs.

yssharma commented:
@felixcheung : would love to get more review comments to improve the patch.

budde commented on docs/streaming-kinesis-integration.md:
```diff
@@ -216,3 +216,7 @@ de-aggregate records during consumption.
   - `InitialPositionInStream.LATEST` could lead to missed records if data is added to the stream while no input DStreams are running (and no checkpoint info is being stored).
   - `InitialPositionInStream.TRIM_HORIZON` may lead to duplicate processing of records where the impact is dependent on checkpoint frequency and processing idempotency.
+
+#### Kinesis retry configurations
```
*nit:* think "configuration" sounds more natural here than "configurations"

budde reviewed (approved):
One tiny nitpick for feedback that you can take or leave but LGTM. I'm satisfied unless @brkyvz or @felixcheung identify any issues. Thanks for iterating on this!

brkyvz commented on docs/streaming-kinesis-integration.md:
```diff
@@ -216,3 +216,7 @@ de-aggregate records during consumption.
   - `InitialPositionInStream.TRIM_HORIZON` may lead to duplicate processing of records where the impact is dependent on checkpoint frequency and processing idempotency.
+
+#### Kinesis retry configurations
+ - `spark.streaming.kinesis.retry.waitTime` : SparkConf for wait time between Kinesis retries (in milliseconds). Default is "100ms".
```
`SparkConf for` is redundant. I would try to focus on when should people actually tweak these, why are these confs important in the first place

brkyvz commented on external/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisBackedBlockRDDSuite.scala:
```diff
@@ -101,6 +103,36 @@ abstract class KinesisBackedBlockRDDTests(aggregateTestData: Boolean)
     }
   }
 
+  testIfEnabled("Basic reading from Kinesis with modified configurations") {
```
I don't see how this test actually tests the configuration setting. It just tests if things work, not that the configurations are actually picked up.

yssharma replied:
I wasn't able to test the actual waiting of Kinesis. I haven't looked at the `PrivateMethodTester ` yet and check how that can help us to test how the vars are picked.
I used this testcase to debug and verify that all the values are passed correctly.

brkyvz commented on external/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisBackedBlockRDDSuite.scala:
```diff
@@ -101,6 +103,36 @@ abstract class KinesisBackedBlockRDDTests(aggregateTestData: Boolean)
 
+  testIfEnabled("Basic reading from Kinesis with modified configurations") {
+    // Add Kinesis retry configurations
+    sc.conf.set(RETRY_WAIT_TIME_KEY, "1000ms")
```
we need to clean these up after the test

yssharma replied:
+1

brkyvz reviewed (changes requested):
Sorry for not providing feedback sooner. There are several things we need to address:
 1. Better documentation. What are these confs aiming to fix
 2. Passing in `sparkConf` is a bit much. Either we should pass in individual confs or use a special case class to pass the configurations
 3. We need a better test to actually test that the configurations are being passed in and used. One option is to use `PrivateMethodTester` to access the internal variables of `KinesisSeqRangeIterator` and make sure they are initialized properly

yssharma commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisBackedBlockRDD.scala:
```diff
@@ -135,7 +139,8 @@ class KinesisSequenceRangeIterator(
     range: SequenceNumberRange,
-    retryTimeoutMs: Int) extends NextIterator[Record] with Logging {
+    retryTimeoutMs: Int,
+    sparkConf: SparkConf) extends NextIterator[Record] with Logging {
```
@brkyvz - I was thinking not to pass individual configs to the constructor because that would just cause the list to grow. Using SparkConf or a Map would enable us to add new configs without any code changes. I was using a Map earlier for this so that its easy to pass more configs. 
What are your thoughts on Map vs Case class ?

brkyvz replied:
I would prefer a specialized case class,
something like:
```scala
case class KinesisReadConfigurations(
  maxRetries: Int,
  retryWaitTimeMs: Long,
  retryTimeoutMs: Long)
```

yssharma replied:
And would you expect it to be passed directly to the`KinesisInputDStream` ?

yssharma replied:
Or we can pass then via spark conf and construct the KinesisReadConfigurations object in `KinesisInputDStream` and pass it down to `KinesisBackedBlockRDD `.

brkyvz replied:
I prefer the latter. Create it in `KinesisInputDStream` and pass it down to `KinesisBackedBlockRDD`

yssharma commented:
@brkyvz - Added new changes that adds -
- A case class `KinesisReadConfigurations` that adds all the kinesis read configs in a single place
- A test class that passes the kinesis configs in `SparkConf` which are then used to create the kinesis configs object in `KinesisInputDStream` and passed down to `KinesisBackedBlockRDD`
- Docs improvement

I also played with the `PrivateMethodTester ` but wasn't able to access the private function `KinesisSequenceRangeIterator#retryOrTimeout` . Probably because of the generics used in the function. I used an alternative to fetch the RDD's directly and check the configs passed in there.
I would still like to learn how to get the `retryOrTimeout` working just out of interest. Adding the error below:

```
    // KinesisSequenceRangeIterator # retryOrTimeout
    val retryOrTimeoutMethod = PrivateMethod[Object]('retryOrTimeout) // <<<- Issue

    val partitions = kinesisRDD.partitions.map {
      _.asInstanceOf[KinesisBackedBlockRDDPartition] }.toSeq

    seqNumRanges1.ranges.map{ range =>
      val seqRangeIter =
        new KinesisSequenceRangeIterator(DefaultCredentials.provider.getCredentials,
        dummyEndpointUrl, dummyRegionName, range, kinesisRDD.kinesisReadConfigs)

      seqRangeIter.invokePrivate(retryOrTimeoutMethod("Passing custom message"))

    }



  - Kinesis read with custom configurations *** FAILED ***
  java.lang.IllegalArgumentException: Can't find a private method named: retryOrTimeout
  at org.scalatest.PrivateMethodTester$Invoker.invokePrivate(PrivateMethodTester.scala:247)
  at org.apache.spark.streaming.kinesis.KinesisStreamTests$$anonfun$7$$anonfun$apply$mcV$sp$13.apply(KinesisStreamSuite.scala:286)
  at org.apache.spark.streaming.kinesis.KinesisStreamTests$$anonfun$7$$anonfun$apply$mcV$sp$13.apply(KinesisStreamSuite.scala:281)
  at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
  at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
  at scala.collection.immutable.List.foreach(List.scala:381)
  at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
  at scala.collection.immutable.List.map(List.scala:285)
  at org.apache.spark.streaming.kinesis.KinesisStreamTests$$anonfun$7.apply$mcV$sp(KinesisStreamSuite.scala:281)
  at org.apache.spark.streaming.kinesis.KinesisStreamTests$$anonfun$7.apply(KinesisStreamSuite.scala:237)
```

yssharma commented:
@budde would you like to share your thoughts on the new changes when you have time ?

brkyvz commented on docs/streaming-kinesis-integration.md:
```diff
@@ -216,3 +216,7 @@ de-aggregate records during consumption.
   - `InitialPositionInStream.TRIM_HORIZON` may lead to duplicate processing of records where the impact is dependent on checkpoint frequency and processing idempotency.
+
+#### Kinesis retry configuration
+ - `spark.streaming.kinesis.retry.waitTime` : Wait time between Kinesis retries as a duration string. When reading from Amazon Kinesis, users may hit 'ThroughputExceededExceptions', when consuming faster than 5 transactions/second or, exceeding the maximum read rate of 2 MB/second. This configuration can be tweaked to increase the sleep between fetches when a fetch fails to reduce these exception ...
```
`ProvisionedThroughputExceededException`s

brkyvz commented on docs/streaming-kinesis-integration.md:
```diff
@@ -216,3 +216,7 @@ de-aggregate records during consumption.
+
+#### Kinesis retry configuration
+ - `spark.streaming.kinesis.retry.waitTime` : Wait time between Kinesis retries as a duration string. When reading from Amazon Kinesis, users may hit 'ThroughputExceededExceptions', when consuming faster than 5 transactions/second or, exceeding the maximum read rate of 2 MB/second. This configuration can be tweaked to increase the sleep between fetches when a fetch fails to reduce these exception ...
+ - `spark.streaming.kinesis.retry.maxAttempts` : Max number of retries for Kinesis fetches. This config can also be used to tackle the Kinesis `ThroughputExceededExceptions` in scenarios mentioned above. It can be increased to have more number of retries for Kinesis reads. Default is 3.
```
ditto

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisBackedBlockRDD.scala:
```diff
@@ -284,17 +282,12 @@ class KinesisSequenceRangeIterator(
         throw new SparkException(
-          s"Timed out after $retryTimeoutMs ms while $message, last exception: ", lastError)
+          s"Timed out after ${kinesisReadConfigs.retryTimeoutMs} ms while "
+          + "$message, last exception: ", lastError)
```
`s"$message`

brkyvz replied:
also please move `+` to line above

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReadConfigurations.scala:
```diff
@@ -0,0 +1,55 @@
+ *                             else uses 10000 if invoked directly.
+ */
+private[kinesis]
+case class KinesisReadConfigurations (
```
How about something like this?
```scala
private[kinesis] case class KinesisReadConfigurations(
    maxRetries: Int,
    retryWaitTimeMs: Long,
    retryTimeoutMs: Long)

object KinesisReadConfigurations {
  def apply(): KinesisReadConfigurations = {
    KinesisReadConfigurations(3, 100, 10000)
  }

  def apply(ssc: StreamingContext): KinesisReadConfigurations = {
    KinesisReadConfigurations(
      maxRetries = ssc.sc.getConf.getInt(RETRY_MAX_ATTEMPTS_KEY, DEFAULT_MAX_RETRIES),
      retryWaitTimeMs = JavaUtils.timeStringAsMs(
        ssc.sc.getConf.getString(RETRY_WAIT_TIME_KEY, DEFAULT_RETRY_WAIT_TIME)),
      retryTimeoutMs = ssc.graph.batchDuration.milliseconds)
  }

  val DEFAULT_MAX_RETRIES = 3
  val DEFAULT_RETRY_WAIT_TIME = "100ms"
}
```

brkyvz commented on external/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisStreamSuite.scala:
```diff
@@ -234,6 +235,50 @@ abstract class KinesisStreamTests(aggregateTestData: Boolean) extends KinesisFun
+    assert(kinesisRDD.kinesisReadConfigs.maxRetries === 5)
+    assert(kinesisRDD.kinesisReadConfigs.retryTimeoutMs === batchDuration.milliseconds)
+
+    ssc.sc.conf.remove(RETRY_WAIT_TIME_KEY)
```
mind putting these in a `try - finally` block

brkyvz commented:
@yssharma left some comments

yssharma commented:
@brkyvz - Thanks for the review comments. Updated the patch, please review.

budde commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReadConfigurations.scala:
```diff
@@ -0,0 +1,71 @@
+ * @param retryWaitTimeMs: The interval between consequent Kinesis retries.
+ *                         Defaults to 100ms.
+ * @param retryTimeoutMs: The timeout in milliseconds for a Kinesis request.
+*                         Defaults to batch duration provided for streaming,
```
*nit*: You're missing a space here and on the following line

budde commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReadConfigurations.scala:
```diff
@@ -0,0 +1,71 @@
+*                         else uses 10000 if invoked directly.
+ */
+private[kinesis] case class KinesisReadConfigurations(
+                                                     maxRetries: Int,
```
Incorrect indentation here-- should be 2 softabs/4 spaces

brkyvz replied:
+1

budde commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReadConfigurations.scala:
```diff
@@ -0,0 +1,71 @@
+
+object KinesisReadConfigurations {
+  def apply(): KinesisReadConfigurations = {
+    KinesisReadConfigurations(3, 100, 10000)
```
I would use constants and named parameters here too, e.g.

```scala
def apply(): KinesisReadConfigurations = KinesisReadConfigurations(
  maxTretries = DEFAULT_MAX_RETRIES,
  ...
```

brkyvz replied:
+1

budde commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReadConfigurations.scala:
```diff
@@ -0,0 +1,71 @@
+import org.apache.spark.streaming.StreamingContext
+
+/**
+ * Configurations to pass to the [KinesisBackedBlockRDD].
```
*nit*: should be ```[[KinesisBackedBlockRDD]]```

brkyvz replied:
+1

budde reviewed (commented):
Took a look and left a few comments, mostly around some style and formatting issues. On the whole, I'm not sure how I feel about introducing the ```KinesisReadConfigurations``` case class if it's just being used to pull a couple config values out from ```SparkConf``` and pass them to KinesisBackedBlockRDD which itself is already taking a ```SparkContext``` as an argument. Reading the config values in the constructor and storing them as ```val``` fields should resolve the serialization issues you were seeing while avoiding having to serialize the entire ```SparkConf```. I'll defer to @brkyvz on this issue though.

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReadConfigurations.scala:
```diff
@@ -0,0 +1,71 @@
+                                                     retryTimeoutMs: Long)
+
+object KinesisReadConfigurations {
+  def apply(): KinesisReadConfigurations = {
```
actually do we even need this?

yssharma replied:
It can be used in places where we don't have the spark conf. I am using this in `KinesisBackedBlockRDD`'s constructor.

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReadConfigurations.scala:
```diff
@@ -0,0 +1,71 @@
+                                                     retryWaitTimeMs: Long,
+                                                     retryTimeoutMs: Long)
+
+object KinesisReadConfigurations {
```
`private object`

brkyvz commented on external/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisStreamSuite.scala:
```diff
@@ -234,6 +235,53 @@ abstract class KinesisStreamTests(aggregateTestData: Boolean) extends KinesisFun
+      assert(kinesisRDD.kinesisReadConfigs.maxRetries === 5)
+      assert(kinesisRDD.kinesisReadConfigs.retryTimeoutMs === batchDuration.milliseconds)
+
+      ssc.sc.conf.remove(RETRY_WAIT_TIME_KEY)
```
these also need to be in the `finally`

brkyvz reviewed (commented):
Looking much better now. Left a couple more comments

yssharma commented:
Thanks for the comments @budde @brkyvz . Would be adding the changes soon.
I too liked pulling the values of spark conf directly and got it working with the private val in `KinesisBackedBlockRDD` [1]. I don't mind getting the conf from the case class either since it keeps all the configs in a place and the class acts as a self documented code. Open to thoughts from you both.

1. https://github.com/yssharma/spark/commit/f5026b4fb1bb98a0d31cfaa6571eee896051aa2b

brkyvz commented:
I really prefer the case class. It's used in places where we don't pass in `SparkContext` as well

budde commented:
Fair enough. I took another look and I think I may have been thinking of the way things worked in an earlier revision of this code. I think the case class is reasonable.

budde commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisInputDStream.scala:
```diff
@@ -60,12 +61,19 @@ private[kinesis] class KinesisInputDStream[T: ClassTag](
         messageHandler = messageHandler,
-        kinesisCreds = kinesisCreds)
+        kinesisCreds = kinesisCreds,
+        kinesisReadConfigs = kinesisReadConfigs)
```
I think it would be sufficient to change this to

```scala
  kinesisReadConfigs = KinesisReadConfigurations(ssc))
```

and omit lines 65-70. I don't think a comment is necessary here, the code is pretty straightforward.

brkyvz replied:
+1

yashs360 committed:
remove kinesis hardcoded retries and pass kinesis read configs

yssharma commented:
Thanks for all the review comments  @budde @brkyvz . Added new review changes.

yssharma commented:
@budde @brkyvz - Any feed back on this one please ?

budde reviewed (approved):
LGTM, apologies for the delay

brkyvz commented:
LGTM! Merging to master/branch-2.2

yssharma commented:
Awesome. Thanks @budde @brkyvz for reviews and patch improvements.
